gax/src/streamingCalls/streaming.ts

```diff
@@ -113,10 +128,189 @@ export class StreamProxy extends duplexify implements GRPCCallResult {
     }
   }
 
+  retry(stream: CancellableStream, retry: RetryOptions) {
+    let retryArgument = this.argument! as unknown as RequestType;
+
+    if (
+      typeof retry.getResumptionRequestFn! === 'function' &&
+      retry.getResumptionRequestFn(retryArgument)
+    ) {
+      retryArgument = retry.getResumptionRequestFn(retryArgument);
+    }
+
+    this.resetStreams(stream);
+
+    const new_stream = this.apiCall!(
+      retryArgument,
+      this._callback
+    ) as CancellableStream;
+    this.stream = new_stream;
+
+    const retryStream = this.streamHandoffHelper(new_stream, retry);
+    if (retryStream !== undefined) {
+      return retryStream;
+    }
+    return new_stream;
+  }
+
+  /**
+   * Helper function to handle total timeout + max retry check for server streaming retries
+   * @param {number} deadline - the current retry deadline
+   * @param {number} maxRetries - maximum total number of retries
+   * @param {number} totalTimeoutMillis - total timeout in milliseconds
+   */
+  timeoutAndMaxRetryCheck(
```

Comment:
nit(naming): it's unclear what the function does from its name. Maybe something like `throwIfWhateverCondition`?